Fast version deangelis for test

katjaweigel opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

    While the checklist is intended to be filled in by the technical and scientific
    reviewers, it is the responsibility of the author of the pull request to make
    sure all items on it are properly implemented.
-->

## Description
Adding testing/recipe_deangelis15nat_fig1_fast.yml, short test recipe to reproduce figure 1 from recipe_deangelis15nat.yml using only one model. It should run around 1 minute on one CPU (Levante). The original recipe failed several times for new releases due to its derrived variables. This short version did run for ESMValCore v2.5 and the current release candidate v2.6.0rc3.
<!--
    Please describe your changes here, especially focusing on why this pull request makes
    ESMValTool better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/en/latest/community/

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->
- Contributes to close [#1636](https://github.com/ESMValGroup/ESMValCore/issues/1636)
- Link to recipe output: https://esmvaltool.dkrz.de/shared/esmvaltool/esmvalbot-output/fast_version_deangelis_for_test-l4mw8nyh/recipe_deangelis15nat_fig1_fast_20220620_153928/
- Link to documentation: https://docs.esmvaltool.org/en/latest/recipes/recipe_deangelis15nat.html
This is the documentation page to the original recipe recipe_deangelis15nat.yml, do the test recipes need an own docu?
* * *

## Before you get started

<!--
    Please discuss your idea with the development team before getting started,
    to avoid disappointment or unnecessary work later. The way to do this is
    to open a new issue on GitHub.
-->

- [x] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [x] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [x] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [x] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [x] [🧪][2] [Recipe is well documented](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recipe-and-diagnostic-documentation)
- [x] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [x] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

bouweandela commented:
@esmvalbot Please run recipe/testing/recipe_deangelis15nat_fig1_fast.yml

bouweandela commented:
@esmvalbot Please run testing/recipe_deangelis15nat_fig1_fast.yml

bouweandela commented on esmvaltool/recipes/testing/recipe_deangelis15nat_fig1_fast.yml:
```diff
@@ -0,0 +1,63 @@
+        project: CMIP5
+        ensemble: r1i1p1
+        mip: Amon
+        field: T2Ms
```
```suggestion
```
`field` is no longer required

katjaweigel replied:
OK, thanks, I will remove it.

bouweandela commented on esmvaltool/recipes/testing/recipe_deangelis15nat_fig1_fast.yml:
```diff
@@ -0,0 +1,63 @@
+      - {dataset: GFDL-CM3, exp: rcp85, start_year: 2091,
+         end_year: 2091}
+      - {dataset: GFDL-CM3, exp: abrupt4xCO2, start_year: 141, end_year: 141}
+        # - {dataset: ACCESS1-0, exp: piControl, start_year: 300, end_year: 301}
+        # - {dataset: ACCESS1-0, exp: rcp85, start_year: 2091, end_year: 2092}
+        # - {dataset: ACCESS1-0, exp: abrupt4xCO2, start_year: 300, end_year: 301}
```
Why are these commented out but not deleted?

katjaweigel replied:
> Thanks! I noticed that there was no provenance for the figure, so I added it for the diagnostic used in this recipe (it looks like it's also missing for the other diagnostic scripts in the same folder though, maybe this would be something to look into).

The provenance should be there, but was so far written in work not in plot.
Should it be in plot now?

katjaweigel replied:
This change:
    - fig.savefig(get_plot_filename('bar_all', cfg), dpi=300)
    + plot_file = get_plot_filename('bar_all', cfg)
    + fig.savefig(plot_file, dpi=300)
    
    Would add an extra variable name. My main issue with the code style checks is always, that it complains about too many variabel names -  therefore I don't like it.

katjaweigel replied:
> Why are these commented out but not deleted?

Sorry, I forgot to delete it (I like the version with two models more, but it is not reliably under 5 minutes).

bouweandela replied:
> Should it be in plot now?

Provenance should be provided for all output files, so both for the files in 'work` and `plot`. This is also written in the documentation: https://docs.esmvaltool.org/en/latest/community/diagnostic.html#provenance-items-provided-by-the-diagnostic-script

bouweandela replied:
> I like the version with two models more, but it is not reliably under 5 minutes

Is there a lot of variation in the runtime? I think we don't need to bee too strict on the 5 minutes, especially now that we only have a few trsting recipes.

katjaweigel replied:
> > Should it be in plot now?
> 
> Provenance should be provided for all output files, so both for the files in `work` and `plot`. This is also written in the documentation: https://docs.esmvaltool.org/en/latest/community/diagnostic.html#provenance-items-provided-by-the-diagnostic-script

In the doku it says:

> with ProvenanceLogger(cfg) as provenance_logger:
>       provenance_logger.log(diagnostic_file, provenance_record)
> The diagnostic_file can be obtained using [esmvaltool.diag_scripts.shared.get_diagnostic_filename](https://docs.esmvaltool.org/en/latest/api/esmvaltool.diag_scripts.shared.html#esmvaltool.diag_scripts.shared.get_diagnostic_filename).

This would write it in work. In the example which is linked there, it is probably written in plot, but it is hard to follow, because I'd first need to follow functions from shared (save_figure) which probably did not even exists, when diagnostic was written. 
That it is really hard to follow and a really good example of things, which are not really backward compatibel and make it hard for people to contribute to the Tool, because even such a small thing like shortening an existing diagnistic for testing purposes creates a lot of extra work :(

katjaweigel replied:
> > I like the version with two models more, but it is not reliably under 5 minutes
> 
> Is there a lot of variation in the runtime? I think we don't need to bee too strict on the 5 minutes, especially now that we only have a few testing recipes. The last run by the esmvalbot only took 20 seconds.

Yes, but I think this could be a special feature of Levantes shared queue. With two models (and 3 experiments each) I got up to nearly 7 minutes run time.

katjaweigel replied:
> > > Should it be in plot now?
> > 
> > 
> > Provenance should be provided for all output files, so both for the files in `work` and `plot`. This is also written in the documentation: https://docs.esmvaltool.org/en/latest/community/diagnostic.html#provenance-items-provided-by-the-diagnostic-script
> 
> In the doku it says:
> 
> > with ProvenanceLogger(cfg) as provenance_logger:
> > provenance_logger.log(diagnostic_file, provenance_record)
> > The diagnostic_file can be obtained using [esmvaltool.diag_scripts.shared.get_diagnostic_filename](https://docs.esmvaltool.org/en/latest/api/esmvaltool.diag_scripts.shared.html#esmvaltool.diag_scripts.shared.get_diagnostic_filename).
> 
> This would write it in work. In the example which is linked there, it is probably written in plot, but it is hard to follow, because I'd first need to follow functions from shared (save_figure) which probably did not even exists, when diagnostic was written. That it is really hard to follow and a really good example of things, which are not really backward compatible and make it hard for people to contribute to the Tool, because even such a small thing like shortening an existing diagnistic for testing purposes creates a lot of extra work :(

@bouweandela The version included now produces 8 files for one plot (*.png + 3 provenance files in plot, *.nc + 3 provenance files in work), while this is not really an issue for this small example, I think it its not a good idea for diagnostics with many figues. Some of the other deangelis diagnostics and some of the martin2018 diagnostics produce for example [Number of Variables] * [Number of Models + 1] plots (e.g. map plots for each model and the multi model mean for several variables). This can easily result in 50-100 plots, if there are 6 additional provenance files that are really too many (and from the code I'd assume the provenance files in work and plot are identical?).

katjaweigel replied:
> > > I like the version with two models more, but it is not reliably under 5 minutes
> > 
> > 
> > Is there a lot of variation in the runtime? I think we don't need to bee too strict on the 5 minutes, especially now that we only have a few testing recipes. The last run by the esmvalbot only took 20 seconds.
> 
> Yes, but I think this could be a special feature of Levantes shared queue. With two models (and 3 experiments each) I got up to nearly 7 minutes run time.

I added the second model again, at the moment it is around 4 minutes for that on one CPU, shared queue on Levante  (but the times there really seem to vary).

bouweandela replied:
> This would write it in work. In the example which is linked there

Indeed, the example is for writing to work, but it also says: "For each output file produced by the diagnostic script, ..." and with that I meant that it's also needed for other files, but I guess that is not clear enough..

bouweandela reviewed (changes requested):
Thanks! I noticed that there was no provenance for the figure, so I added it for the diagnostic used in this recipe (it looks like it's also missing for the other diagnostic scripts in the same folder though, maybe this would be something to look into).

> This is the documentation page to the original recipe recipe_deangelis15nat.yml, do the test recipes need an own docu?

Yes, I think so. No need for scientific documentation, but it would be good to list the testing recipes and their options and for each recipe why it is considered a good test. Could you add a documentation page for the testing recipes? A single documentation page for all recipes in the `testing` directory would probably be most convenient.

katjaweigel commented:
> Thanks! I noticed that there was no provenance for the figure, so I added it for the diagnostic used in this recipe (it looks like it's also missing for the other diagnostic scripts in the same folder though, maybe this would be something to look into).
> 
> > This is the documentation page to the original recipe recipe_deangelis15nat.yml, do the test recipes need an own docu?
> 
> Yes, I think so. No need for scientific documentation, but it would be good to list the testing recipes and their options (or link to the original recipe page) and for each recipe why it is considered a good test. Could you add a documentation page for the testing recipes? A single documentation page for all recipes in the `testing` directory would probably be most convenient.

I added a docu page, where also the other test recipes could be added: doc/sphinx/source/recipes/recipes_testing.rst
and put it under "Others", is it ok like this? (Does the link to the docu for recipe_deangelis15nat.yml version work?)

bouweandela commented:
>  because even such a small thing like shortening an existing diagnistic for testing purposes creates a lot of extra work

What extra work are you are talking about? 3 lines of code needed to be added to fix the provenance for this recipe and I did that for you, so where is the extra work for you?

>  This can easily result in 50-100 plots, if there are 6 additional provenance files that are really too many (and from the code I'd assume the provenance files in work and plot are identical?).

In the case of this diagnostic yes, but not all diagnostics create a plot and NetCDF file with the same data. Some only create plots or even other file formats. In any case, the feature that creates the 2 extra citation info files was added over 2 years ago in ESMValGroup/ESMValCore#402, so it is a bit late to complain about the number of output files now, but you're welcome to open an issue if you have ideas about how this could be improved. I do agree that it are many files and it would look a bit messy if you try to open this with a file browser. If you view the results by opening the index.html file in your browser the presentation is nicer.

> Does the link to the docu for recipe_deangelis15nat.yml version work?

I fixed it for you, you needed to add backticks around the link name. You can preview the documentation by clicking on the docs/readthedocs.org:esmvaltool check below.

bouweandela commented:
Thanks! This gives me a good start to implement some automated recipe testing.

katjaweigel commented:
> > because even such a small thing like shortening an existing diagnistic for testing purposes creates a lot of extra work
> 
> What extra work are you are talking about? 3 lines of code needed to be added to fix the provenance for this recipe and I did that for you, so where is the extra work for you?
> 
> > This can easily result in 50-100 plots, if there are 6 additional provenance files that are really too many (and from the code I'd assume the provenance files in work and plot are identical?).
> 
> In the case of this diagnostic yes, but not all diagnostics create a plot and NetCDF file with the same data. Some only create plots or even other file formats. In any case, the feature that creates the 2 extra citation info files was added over 2 years ago in [ESMValGroup/ESMValCore#402](https://github.com/ESMValGroup/ESMValCore/pull/402), so it is a bit late to complain about the number of output files now, but you're welcome to open an issue if you have ideas about how this could be improved. I do agree that it are many files and it would look a bit messy if you try to open this with a file browser. If you view the results by opening the index.html file in your browser the presentation is nicer.
> 
> > Does the link to the docu for recipe_deangelis15nat.yml version work?
> 
> I fixed it for you, you needed to add backticks around the link name. You can preview the documentation by clicking on the docs/readthedocs.org:esmvaltool check below.

Thanks for fixing the link and merging!
With extra work I meat first understanding the issue, but also, that I now should ideally start new PRs for my more then 2 year old recipes to fix that  - leading back to the endless discussion about who has to adapt existing recipes to new technical requirements.

bouweandela commented:
Adding provenance is not a new technical requirement, it has been around since January 2019, long before there was a stable release and the promise of backward compatibility. Anyway, let's not argue on GitHub, it would probably be better to discuss this in person at the next workshop.

katjaweigel commented:
> Adding provenance is not a new technical requirement, it has been around since January 2019, long before there was a stable release and the promise of backward compatibility. Anyway, let's not argue on GitHub, it would probably be better to discuss this in person at the next workshop.

The provenace was there, only not in the correct folder (although it is not really clear, which is the correct folder follwing the doku text). I really see the point to require provanance in a standardtised way, but from the doku/ PR you linked I don't really understand, that the standard should be 6 provenance files in two different directories for every plot...

bouweandela commented:
Yes, maybe we could look at the documentation together at the next workshop and make it more clear?

katjaweigel commented:
In principle yes, but I'm afraid any change would lead either to the request to change something in many diagnostics or there will continue to be a lot of different solutions depending on who wrote a diagnostic when.
